doc/auth_design.md

```diff
@@ -0,0 +1,113 @@
+# Design: SQLFlow Authentication and Authorization
+
+## Terminology and Background
+
+For a quick explanation, the word "Authentication" means to identify the
+user recognized by the system is the actual real-world user. And "Authorization"
+means to grant privileges to the user to access some part of the system
+functionalities.
+
+SQLFlow works as a "bridge" between databases and
+Deep Learning/Machine Learning frameworks. In order to execute a job,
+SQLFlow need both permissions to access databases and submit jobs to
+systems to run distributed training jobs, like submitting jobs to Kubernetes
+to run a distributed tensorflow job.
+
+<img src="figures/auth1.png">
+
+In production environments, one SQLFlow server is designed to accept many clients'
+connections and job submissions. In this case, we must securely store a mapping
+from the user's ID to the user's credentials for accessing both the database and the
+training cluster. Then server-side "session" should also be considered during
+implementing authentication.
+
+For authorization, it will be definitely reasonable to directly proxy requests to
+databases and training clusters, the request will be denied if the current user
+have no access to the requested service.
+
+## Design
+
+An authentication server (for short, will use "auth server" instead) will be introduced
+to achieve extensible authentication configurations. We use a
+[Django](https://www.djangoproject.com/) web server so that the authentication methods
+can extend to:
+
+- Database authentication
+- LDAP
+- User-defined authentication methods
+
+### Session
+
+A server-side "session" is needed to store credentials for each client to access
+the database and submitting jobs. The session can be defined as:
+
+```go
+type Session struct {
+    ClientEndpoint    string
+    DBConnStr         string  // mysql://user:pass@127.0.0.1:3306
+    Token             int64  // useful only in "side-car" design
+}
```

Comment:
Should we consider if a session expired?

Reply:
We should! Thanks!